### ISSUE TITLE
Enable Wayland support for SDL2 in upcoming runtime releases

### ISSUE THREAD

**Issue by kon14** [ValveSoftware/steam-runtime#487]
With SDL finally [switching its default backend to Wayland](https://twitter.com/icculus/status/1486393168211976193) would you consider building SDL2 with Wayland support so that more games can run natively instead of going through XWayland?

Most of the popular distros already default to Wayland for mesa users on their Gnome and KDE spins and should soon start defaulting to it for Nvidia too (eg Fedora 36).
Even if you choose to override the default for a while longer, it would be nice if users could at the very least explicitly use `SDL_VIDEODRIVER=wayland` themselves without having to preload their host libs, which shouldn't even be possible in pressure-vessel due to bubblewrap(?)

**Comment by kisak-valve** [comment]
Hello @kon14, please see https://github.com/ValveSoftware/steam-runtime/issues/396#issuecomment-823465633 and https://github.com/ValveSoftware/steam-runtime/issues/473#issuecomment-953975321.

**Comment by kon14** [comment]
I was just about to edit my OP regarding this.
I get that this is not feasible on scout, thus the "upcoming" part in the title.

I'm not particularly up to date with what release soldier is based on, but your second link mentions SDL2's build on it does not disable the Wayland backend. Is there any way to know what runtime a game is launched with depending on whether it's a native title, a native title running in pressure-vessel using SteamLinuxRuntime or a >=5.13 Proton?

**Comment by smcv** [comment]
> it would be nice if users could at the very least explicitly use `SDL_VIDEODRIVER=wayland` themselves

The tl;dr version is: you can usually already do that, but I wouldn't recommend it yet; and if it doesn't work, it isn't necessarily the Steam Runtime's fault.

Until ValveSoftware/steam-for-linux#8020 is resolved, native Wayland games will have various problems with the Steam Overlay, which could manifest as either crashing, or being unable to access Steam Overlay functionality (including game controllers that are mediated by Steam Input, and emulated input from other machines using Steam Link / Remote Play / Remote Play Together). The Steam Runtime cannot solve this, because the Steam Overlay is one of the proprietary parts of Steam (which is why it's a steam-for-linux issue and not a steam-runtime issue).

## soldier

> I'm not particularly up to date with what release soldier is based on

It's based on Debian 10, but we backport various packages, including SDL 2.

SDL 2 in soldier is already meant to have the Wayland backend enabled. libdecor (window decorations for non-fullscreen windows) is *not* enabled yet: we can't include that in soldier unless we backport a newer libwayland, which is a stability risk (it might still happen, but it isn't obviously easy and correct).

If you find that the Wayland backend is not, in fact, enabled in soldier, please report that as an issue - it's meant to be.

## sniper

The next Steam Runtime is [sniper](https://repo.steampowered.com/steamrt-images-sniper/snapshots/), based on Debian 11, which is new enough that backporting libdecor was straightforward; so the initial public release of sniper already has libdecor, and support for it is enabled in SDL.

sniper isn't used for anything right now, but I'm hoping that it will become possible in future for Steam games to opt-in to using sniper instead of scout, in which case they will automatically use a Wayland-enabled SDL all the time.

## scout

> I get that this is not feasible on scout, thus the "upcoming" part in the title.

Actually, it sort of is...

For native Linux games running in the traditional scout LD_LIBRARY_PATH runtime, if you have a system copy of SDL 2 that is the same word size as the game, and its version is equal to or newer than the Steam Runtime's version, we'll use your system copy in preference to ours. For 32-bit games, remember that you will need a 32-bit SDL 2, which is probably not installed by default: for example, that's `libsdl2-2.0-0:i386` on Debian/Ubuntu or `lib32-sdl2` on Arch.

For native Linux games that you have specifically configured to use the `Steam Linux Runtime` compatibility tool, we're using a mixture of scout and soldier, preferring soldier for cases where the same library exists in both places - so you should get the same SDL that soldier has, which as I said above has Wayland enabled, but not libdecor at the moment.

## Detecting which one you have

> Is there any way to know what runtime a game is launched with depending on whether it's a native title, a native title running in pressure-vessel using SteamLinuxRuntime or a Windows build running with >=5.13 Proton?

While the game is running, if `pstree` or `systemd-cgls` or similar tools finds programs prefixed with `pressure-vessel-` or `pv-`, then you're using pressure-vessel, which currently always means soldier's SDL2 version. If not, you're using the old scout runtime.

**Comment by kon14** [comment]
Portal2 (and mods) seem to work. I came across this a while ago while attempting this, but Dota2 and most other SDL/Source games just wouldn't. I'll test this further, forcing pressure-vessel.

Dota2 for instance flatout refuses to work even though it used to at some point in the past.
Ever since its latest update it uses pressure-vessel ootb (confirmed with pstree) so forcing it shouldn't be necessary, but I tried that also nonetheless. I also specified -gl as there had been rumors of Vulkan being broken on the Wayland backend.

**Comment by smcv** [comment]
Ah, now I know what's happening there. Dota 2 bundles its own copy of SDL 2, overriding the one from the Steam Runtime (either scout or a container)... so the Steam Runtime is not in control of whether Dota 2 supports Wayland or not.

Portal 2 is an example of a game that behaves the way the Steam Runtime is designed to work, with no included libraries that duplicate what the Steam Runtime provides. Dota 2, not so much.

That's a Dota 2 issue, already opened as https://github.com/ValveSoftware/Dota-2/issues/999, which Dota 2 developers could either resolve by removing their copy of SDL 2 (which would mean it picks up whatever the Steam Runtime and/or the OS give it), or by building *their* copy of SDL 2 with Wayland support.

**Comment by kon14** [comment]
Seems like that's indeed the case. Overriding with own SDL gets it launching natively:
`SDL_DYNAMIC_API=/lib/x86_64-linux-gnu/libSDL2-2.0.so.0 SDL_VIDEODRIVER=wayland %command%`

It seems like using the OpenGL renderer is not necessary anymore, though my perf is awful, most likely due to some Nvidia-specific bug, so can't really tell. It's offtopic either way.

Thanks for being so informative, it's really appreciated.
Hopefully less games will depend on XWayland in the not so distant future.

I'm not really sure if I should close the issue at this point.
Leaving it on, but feel free to resolve if you feel like it's as done as it gets.

**Comment by smcv** [comment]
> I'm not really sure if I should close the issue at this point

I think we should probably close it . There will hopefully be some incremental improvements in soldier (in particular I want to look at whether we can enable libdecor) but there isn't actually anything there that is blocking use of Wayland, so having this issue open is a bit misleading.

**Comment by kisak-valve** [comment]
Closing per the last couple comments.

**Comment by Ristovski** [comment]
Just a PSA: I got TF2 running natively on Wayland by going to the game properties and setting the following under `Launch options`: `LD_PRELOAD=/app/lib/i386-linux-gnu/libSDL2-2.0.so.0.2400.0 SDL_VIDEODRIVER=wayland %command%`.

This uses the SDL2 lib thats part of `org.freedesktop.Platform` which comes as a dependency of Steam (for flatpak).

Note the `i386-linux-gnu` - you might need to change this for 64bit titles.

**Comment by smcv** [comment]
If a game bundles its own copy of SDL, then overriding that with a different version is not a supported thing to do. Ideally, most (all?) games on Steam should stop providing their own bundled copy of SDL, so that their game will use the Steam Runtime's version (which is regularly updated). However, the maintainers of the Steam Runtime cannot force game developers to do this: it's up to the game developer. The issue tracker for TF2 seems to be https://github.com/ValveSoftware/Source-1-Games/issues.

Forcing a different version of a library via `LD_PRELOAD` is not really a supportable thing to do in general: it forces the requested library into *all* programs that are run as part of launching TF2 (not just TF2 itself!) which can easily break things, especially if that library was not designed to be `LD_PRELOAD`ed. It's sometimes a useful workaround, but only use it if you are able to debug what happens when it goes wrong.

For SDL specifically, a more narrowly-scoped workaround than `LD_PRELOAD` is to [set `SDL_DYNAMIC_API` to the path to a suitable SDL version](https://github.com/libsdl-org/SDL/blob/main/docs/README-dynapi.md), which only affects programs that are already linked to SDL.

You should also be aware that the SDL library in `org.freedesktop.Platform` (or on your host system, for non-Flatpak users) might be either older or newer than the SDL in the Steam Runtime (`$STEAM_RUNTIME/usr/lib/i386-linux-gnu/libSDL2-2.0.so.0`), because the SDL in the Steam Runtime is regularly updated. The Steam Runtime's SDL is currently 2.24.0 for most users, or 2.24.1 if you are using the Steam client beta. If you force the use of a version of SDL that is older than the one the game expects to use, it is very likely to crash.

Whatever version of SDL you're using, I would not recommend forcing use of `SDL_VIDEODRIVER=wayland` until the Steam Overlay [supports it](https://github.com/ValveSoftware/steam-for-linux/issues/8020).
